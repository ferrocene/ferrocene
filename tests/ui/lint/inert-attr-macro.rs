//@ check-pass

#![feature(rustc_attrs)]
#![warn(unused)]

macro_rules! foo {
    () => {}
}

fn main() {
    #[rustc_dummy] foo!(); //~ WARN unused attribute `rustc_dummy`

    // This does nothing, since `#[allow(warnings)]` is itself
    // an inert attribute on a macro call
    #[allow(warnings)] #[rustc_dummy] foo!(); //~ WARN unused attribute `allow`
    //~^ WARN unused attribute `rustc_dummy`

    // This does work, since the attribute is on a parent
    // of the macro invocation.
    #[allow(warnings)] { #[rustc_dummy] foo!(); }

    // Ok, `cfg` and `cfg_attr` are expanded eagerly and do not warn.
    #[cfg(true)] foo!();
    #[cfg(false)] foo!();
    #[cfg_attr(true, cfg(true))] foo!();
    #[cfg_attr(false, nonexistent)] foo!();
}
