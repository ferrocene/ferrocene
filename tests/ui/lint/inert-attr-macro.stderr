warning: unused attribute `rustc_dummy`
  --> $DIR/inert-attr-macro.rs:11:5
   |
LL |     #[rustc_dummy] foo!();
   |     ^^^^^^^^^^^^^^
   |
note: the built-in attribute `rustc_dummy` will be ignored, since it's applied to the macro invocation `foo`
  --> $DIR/inert-attr-macro.rs:11:20
   |
LL |     #[rustc_dummy] foo!();
   |                    ^^^
note: the lint level is defined here
  --> $DIR/inert-attr-macro.rs:4:9
   |
LL | #![warn(unused)]
   |         ^^^^^^
   = note: `#[warn(unused_attributes)]` implied by `#[warn(unused)]`

warning: unused attribute `allow`
  --> $DIR/inert-attr-macro.rs:15:5
   |
LL |     #[allow(warnings)] #[rustc_dummy] foo!();
   |     ^^^^^^^^^^^^^^^^^^
   |
note: the built-in attribute `allow` will be ignored, since it's applied to the macro invocation `foo`
  --> $DIR/inert-attr-macro.rs:15:39
   |
LL |     #[allow(warnings)] #[rustc_dummy] foo!();
   |                                       ^^^

warning: unused attribute `rustc_dummy`
  --> $DIR/inert-attr-macro.rs:15:24
   |
LL |     #[allow(warnings)] #[rustc_dummy] foo!();
   |                        ^^^^^^^^^^^^^^
   |
note: the built-in attribute `rustc_dummy` will be ignored, since it's applied to the macro invocation `foo`
  --> $DIR/inert-attr-macro.rs:15:39
   |
LL |     #[allow(warnings)] #[rustc_dummy] foo!();
   |                                       ^^^

warning: 3 warnings emitted

